File: src/ai/skills/web-search.ts
/**
 * Web Search system skill · Brave Search and/or Tavily API clients.
 *
 * Used by the orchestrator (room.ts) before each agent speaks: when
 * the Pass-1 router decides the turn would benefit from fresh web
 * info, we hit the configured backend for the chosen query, distill
 * the top results into a SHARED MATERIALS block, and prepend it to the
 * agent's Pass-2 system prompt.
 *
 * Failures are non-fatal — a search timeout / 4xx / 5xx degrades to
 * "no search this turn", and the agent answers from its own
 * training without the panel ever showing an error to the user.
 */

const BRAVE_ENDPOINT = "https://api.search.brave.com/res/v1/web/search";
const TAVILY_ENDPOINT = "https://api.tavily.com/search";
const DEFAULT_TIMEOUT_MS = 6000;
const DEFAULT_RESULT_COUNT = 5;

export type WebSearchBackend = "brave" | "tavily";

export interface BraveResult {
  title: string;
  url: string;
  description: string;
  /** ISO 8601 publication date when Brave surfaces one. */
  age?: string;
}

export interface BraveSearchOpts {
  apiKey: string;
  query: string;
  /** Number of results to return (1-10). */
  count?: number;
  /** Country / language hint. ISO codes from Brave docs. */
  country?: string;
  searchLang?: string;
  timeoutMs?: number;
}

/** Run a Brave Search query. Returns up to `count` results, or null
 *  on any error (timeout / 4xx / 5xx / parse failure). The caller is
 *  expected to treat null as "no search this turn" and proceed. */
export async function runBraveSearch(opts: BraveSearchOpts): Promise<BraveResult[] | null> {
  const apiKey = opts.apiKey.trim();
  const query = opts.query.trim();
  if (!apiKey || !query) return null;

  const params = new URLSearchParams({
    q: query,
    count: String(Math.min(Math.max(opts.count ?? DEFAULT_RESULT_COUNT, 1), 10)),
    safesearch: "moderate",
  });
  if (opts.country) params.set("country", opts.country);
  if (opts.searchLang) params.set("search_lang", opts.searchLang);

  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), opts.timeoutMs ?? DEFAULT_TIMEOUT_MS);

  try {
    const res = await fetch(`${BRAVE_ENDPOINT}?${params.toString()}`, {
      method: "GET",
      headers: {
        Accept: "application/json",
        "X-Subscription-Token": apiKey,
      },
      signal: ctrl.signal,
    });
    if (!res.ok) {
      process.stderr.write(`[web-search] brave returned ${res.status}\n`);
      return null;
    }
    const json = await res.json() as { web?: { results?: Array<{
      title?: string;
      url?: string;
      description?: string;
      age?: string;
    }> } };
    const raw = (json && json.web && json.web.results) || [];
    const out: BraveResult[] = [];
    for (const r of raw) {
      const title = (r.title || "").trim();
      const url = (r.url || "").trim();
      if (!title || !url) continue;
      out.push({
        title,
        url,
        description: stripHtml(r.description || "").trim(),
        age: r.age,
      });
      if (out.length >= (opts.count ?? DEFAULT_RESULT_COUNT)) break;
    }
    return out;
  } catch (e) {
    process.stderr.write(
      `[web-search] brave fetch failed: ${e instanceof Error ? e.message : String(e)}\n`,
    );
    return null;
  } finally {
    clearTimeout(timer);
  }
}

export interface TavilySearchOpts {
  apiKey: string;
  query: string;
  maxResults?: number;
  timeoutMs?: number;
}

/** Run a Tavily Search query · same return shape as Brave for prompt formatters. */
export async function runTavilySearch(opts: TavilySearchOpts): Promise<BraveResult[] | null> {
  const apiKey = opts.apiKey.trim();
  const query = opts.query.trim();
  if (!apiKey || !query) return null;

  const maxResults = Math.min(Math.max(opts.maxResults ?? DEFAULT_RESULT_COUNT, 1), 10);
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), opts.timeoutMs ?? DEFAULT_TIMEOUT_MS);

  try {
    const res = await fetch(TAVILY_ENDPOINT, {
      method: "POST",
      headers: { Accept: "application/json", "content-type": "application/json" },
      body: JSON.stringify({
        api_key: apiKey,
        query,
        search_depth: "basic",
        max_results: maxResults,
      }),
      signal: ctrl.signal,
    });
    if (!res.ok) {
      process.stderr.write(`[web-search] tavily returned ${res.status}\n`);
      return null;
    }
    const json = await res.json() as { results?: Array<{
      title?: string;
      url?: string;
      content?: string;
      published_date?: string;
    }> };
    const raw = (json && json.results) || [];
    const out: BraveResult[] = [];
    for (const r of raw) {
      const title = (r.title || "").trim();
      const url = (r.url || "").trim();
      if (!title || !url) continue;
      const desc = stripHtml((r.content || "").replace(/\s+/g, " "));
      out.push({
        title,
        url,
        description: desc,
        age: r.published_date,
      });
      if (out.length >= maxResults) break;
    }
    return out;
  } catch (e) {
    process.stderr.write(
      `[web-search] tavily fetch failed: ${e instanceof Error ? e.message : String(e)}\n`,
    );
    return null;
  } finally {
    clearTimeout(timer);
  }
}

/** Dispatch to Brave or Tavily based on the user's active search backend. */
export async function runWebSearch(
  backend: WebSearchBackend,
  apiKey: string,
  query: string,
  opts?: { count?: number; timeoutMs?: number },
): Promise<BraveResult[] | null> {
  if (backend === "tavily") {
    return runTavilySearch({
      apiKey,
      query,
      maxResults: opts?.count,
      timeoutMs: opts?.timeoutMs,
    });
  }
  return runBraveSearch({
    apiKey,
    query,
    count: opts?.count,
    timeoutMs: opts?.timeoutMs,
  });
}

/** Brave returns descriptions with `<strong>` highlighting. Keep them
 *  as plain text for prompt injection — strip tags + decode the few
 *  entity references that show up. */
function stripHtml(s: string): string {
  return s
    .replace(/<\/?[a-z][^>]*>/gi, "")
    .replace(/&nbsp;/g, " ")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/\s+/g, " ");
}

/** Distill search results into the SHARED MATERIALS block injected
 *  into the agent's Pass-2 system prompt. The format mirrors the
 *  fetch-url skill so agents already know how to cite this. */
export function formatSearchResults(query: string, results: BraveResult[]): string {
  if (!results.length) return "";
  const lines: string[] = [];
  lines.push("─── SHARED MATERIALS · WEB SEARCH ───");
  lines.push("");
  lines.push(`Query: ${query}`);
  lines.push("");
  lines.push(
    "Cite sources by their bracketed number when you use a fact below " +
      "(e.g. \"...as reported in [2]\"). If a source contradicts your " +
      "training, prefer the search result — it's more recent.",
  );
  lines.push("");
  results.forEach((r, i) => {
    const n = i + 1;
    const age = r.age ? ` · ${r.age}` : "";
    lines.push(`[${n}] ${r.title}${age}`);
    lines.push(`    ${r.url}`);
    if (r.description) {
      const desc = r.description.length > 320
        ? r.description.slice(0, 317) + "…"
        : r.description;
      lines.push(`    ${desc}`);
    }
    lines.push("");
  });
  lines.push("─── END SHARED MATERIALS ───");
  return lines.join("\n");
}

File: ios/BoardroomKit/Sources/BoardroomSearch/WebSearch.swift
import Foundation

/// Web search — Brave (GET) / Tavily (POST), ported from `src/ai/skills/web-search.ts`.
/// Failures are NON-fatal: any timeout / 4xx / 5xx / parse error returns nil and
/// the caller proceeds without search (the agent answers from training). 6s
/// timeout. `formatResults` renders the SHARED MATERIALS prompt block.
public struct SearchResult: Sendable, Equatable {
    public let title: String
    public let url: String
    public let description: String
    public let age: String?      // ISO publication date when surfaced
    public init(title: String, url: String, description: String, age: String?) {
        self.title = title; self.url = url; self.description = description; self.age = age
    }
}

public enum SearchBackend: String, Sendable { case brave, tavily }

public struct WebSearch: Sendable {
    public static let defaultTimeout: TimeInterval = 6
    public static let defaultCount = 5

    let session: URLSession
    public init(session: URLSession = .shared) { self.session = session }

    /// Dispatch to the active backend. Returns nil on any failure.
    public func run(_ backend: SearchBackend, apiKey: String, query: String,
                    count: Int = defaultCount, timeout: TimeInterval = defaultTimeout) async -> [SearchResult]? {
        switch backend {
        case .brave:  return await brave(apiKey: apiKey, query: query, count: count, timeout: timeout)
        case .tavily: return await tavily(apiKey: apiKey, query: query, count: count, timeout: timeout)
        }
    }

    private func brave(apiKey: String, query: String, count: Int, timeout: TimeInterval) async -> [SearchResult]? {
        let key = apiKey.trimmingCharacters(in: .whitespaces)
        let q = query.trimmingCharacters(in: .whitespaces)
        guard !key.isEmpty, !q.isEmpty, let req = Self.braveRequest(apiKey: key, query: q, count: count, timeout: timeout)
        else { return nil }
        guard let (data, resp) = try? await session.data(for: req),
              let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { return nil }
        return Self.parseBrave(data, limit: max(1, min(count, 10)))
    }

    private func tavily(apiKey: String, query: String, count: Int, timeout: TimeInterval) async -> [SearchResult]? {
        let key = apiKey.trimmingCharacters(in: .whitespaces)
        let q = query.trimmingCharacters(in: .whitespaces)
        guard !key.isEmpty, !q.isEmpty, let req = Self.tavilyRequest(apiKey: key, query: q, count: count, timeout: timeout)
        else { return nil }
        guard let (data, resp) = try? await session.data(for: req),
              let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { return nil }
        return Self.parseTavily(data, limit: max(1, min(count, 10)))
    }

    // MARK: Pure builders + parsers (unit-tested)

    public static func braveRequest(apiKey: String, query: String, count: Int, timeout: TimeInterval) -> URLRequest? {
        var comps = URLComponents(string: "https://api.search.brave.com/res/v1/web/search")
        comps?.queryItems = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "count", value: String(max(1, min(count, 10)))),
            URLQueryItem(name: "safesearch", value: "moderate"),
        ]
        guard let url = comps?.url else { return nil }
        var req = URLRequest(url: url, timeoutInterval: timeout)
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue(apiKey, forHTTPHeaderField: "X-Subscription-Token")
        return req
    }

    public static func tavilyRequest(apiKey: String, query: String, count: Int, timeout: TimeInterval) -> URLRequest? {
        guard let url = URL(string: "https://api.tavily.com/search") else { return nil }
        let body: [String: Any] = ["api_key": apiKey, "query": query,
                                   "search_depth": "basic", "max_results": max(1, min(count, 10))]
        var req = URLRequest(url: url, timeoutInterval: timeout)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try? JSONSerialization.data(withJSONObject: body, options: [.sortedKeys])
        return req
    }

    public static func parseBrave(_ data: Data, limit: Int) -> [SearchResult] {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let web = json["web"] as? [String: Any],
              let raw = web["results"] as? [[String: Any]] else { return [] }
        var out: [SearchResult] = []
        for r in raw {
            let title = ((r["title"] as? String) ?? "").trimmingCharacters(in: .whitespaces)
            let url = ((r["url"] as? String) ?? "").trimmingCharacters(in: .whitespaces)
            guard !title.isEmpty, !url.isEmpty else { continue }
            out.append(SearchResult(title: title, url: url,
                                    description: stripHtml(r["description"] as? String ?? "").trimmingCharacters(in: .whitespaces),
                                    age: r["age"] as? String))
            if out.count >= limit { break }
        }
        return out
    }

    public static func parseTavily(_ data: Data, limit: Int) -> [SearchResult] {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let raw = json["results"] as? [[String: Any]] else { return [] }
        var out: [SearchResult] = []
        for r in raw {
            let title = ((r["title"] as? String) ?? "").trimmingCharacters(in: .whitespaces)
            let url = ((r["url"] as? String) ?? "").trimmingCharacters(in: .whitespaces)
            guard !title.isEmpty, !url.isEmpty else { continue }
            let content = (r["content"] as? String ?? "").replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            out.append(SearchResult(title: title, url: url, description: stripHtml(content),
                                    age: r["published_date"] as? String))
            if out.count >= limit { break }
        }
        return out
    }

    static func stripHtml(_ s: String) -> String {
        s.replacingOccurrences(of: #"(?i)</?[a-z][^>]*>"#, with: "", options: .regularExpression)
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
    }

    /// SHARED MATERIALS prompt block — mirrors `formatSearchResults` exactly so
    /// the prompt bytes match the desktop.
    public static func formatResults(query: String, results: [SearchResult]) -> String {
        if results.isEmpty { return "" }
        var lines: [String] = ["─── SHARED MATERIALS · WEB SEARCH ───", "", "Query: \(query)", ""]
        lines.append("Cite sources by their bracketed number when you use a fact below "
                   + "(e.g. \"...as reported in [2]\"). If a source contradicts your "
                   + "training, prefer the search result — it's more recent.")
        lines.append("")
        for (i, r) in results.enumerated() {
            let age = r.age.map { " · \($0)" } ?? ""
            lines.append("[\(i + 1)] \(r.title)\(age)")
            lines.append("    \(r.url)")
            if !r.description.isEmpty {
                let desc = r.description.count > 320 ? String(r.description.prefix(317)) + "…" : r.description
                lines.append("    \(desc)")
            }
            lines.append("")
        }
        lines.append("─── END SHARED MATERIALS ───")
        return lines.joined(separator: "\n")
    }
}
